api/include/opentelemetry/logs/noop.h

```diff
@@ -34,11 +34,34 @@ class NoopLogger final : public Logger
 public:
   const nostd::string_view GetName() noexcept override { return "noop logger"; }
 
-  nostd::unique_ptr<LogRecord> CreateLogRecord() noexcept override { return nullptr; }
+  nostd::unique_ptr<LogRecord> CreateLogRecord() noexcept override {
+    return nostd::unique_ptr<LogRecord>(&record_);
+  }
```

Comment:
This uses a member of the NoopLogger itself, which may go away, leaving the record pointer dangling.

Possible improvement:

```suggestion
  nostd::unique_ptr<LogRecord> CreateLogRecord() noexcept override {
    static NoopLogRecord dummy;
  
    return nostd::unique_ptr<LogRecord>(&dummy);
  }
```

The static will never go out of scope.

Reply:
Yes, as long as it is inside a method. A static member in a class will not work.

Trying this at the moment, and still have a runtime failure in the 2 failing tests.

Reply:
From experiments, faking the new and delete is only part of the issue.

When a log record is created, used in a unique_ptr, and when the unique_ptr is destroyed,
the object pointed to is destroyed (destructor) and then memory is reclaimed (delete).

When the destructor is called on the dummy shared object, the virtual table is damaged,
so that reusing the same object for another log record leads to a crash (pure virtual method called).

Trying with a Deleter in the unique_ptr.

Reply:
And of course:

* `std::unique_ptr<LogRecord>`
* `std::unique_ptr<LogRecord, Deleter>`

are two different classes, one can not assign the second to the first, so there is no way to return a unique pointer with a custom deleter for the `NoopLogRecord`.

The conclusion so far is that unique_ptr(& dummy) owns the dummy object, and happily trashes it including the virtual table when the unique pointer goes out of scope. This prevents reusing the same dummy object for different log records.